File: web/apps/app/lib/core/metadata/__tests__/metadata-fixture.test.ts
import 'fake-indexeddb/auto';
/**
 * @vitest-environment jsdom
 *
 * Web-side consumer test for the shared metadata fixture.
 *
 * Seeds the cache (IndexedDB) from `shared/test-fixtures/relay/metadata-fixture.json` and checks
 * that metadata-client parses the fixture's matched samples and relayRuntimeConfig correctly.
 */
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { pruneBlobs } from '../../../infra/storage/blob-cache';
import { __seedMetadataCacheForTest as seedMetadataCache } from '../metadata-client';

// vitest runs with cwd = apps/app, so walk up three levels to the repository root.
const FIXTURE_PATH = resolve(
  process.cwd(),
  '../../..',
  'shared/test-fixtures/relay/metadata-fixture.json',
);

function loadFixture(): Record<string, unknown> {
  const raw = readFileSync(FIXTURE_PATH, 'utf-8');
  return JSON.parse(raw) as Record<string, unknown>;
}

describe('metadata-fixture (shared with iOS / Android)', () => {
  beforeEach(async () => {
    vi.resetModules();
    vi.restoreAllMocks();
    localStorage.clear();
    await pruneBlobs('oriveo:metadata:c', []);
    vi.spyOn(globalThis, 'fetch').mockResolvedValue({
      status: 304,
      ok: false,
      json: vi.fn(),
    } as unknown as Response);
  });

  it('resolves gpt-5.4 / gpt-5.4-2026-04-01 / gpt-image-2 / claude-sonnet-4.5 / gemini-2.5-pro', async () => {
    const fixture = loadFixture();
    await seedMetadataCache({
      timestamp: Date.now(),
      data: fixture,
    });

    const metadata = await import('../metadata-client');
    await metadata.initMetadata();

    // Cross-provider match, in the specified order.
    const gpt = metadata.resolveCatalogModelAcrossProvidersWithProvider('gpt-5.4');
    expect(gpt?.matchedProviderKind).toBe('openAI');
    expect(gpt?.canonicalModelId).toBe('gpt-5.4');
    expect(gpt?.metadata.displayName).toBe('GPT-5.4');

    // Dated snapshot alias match.
    const gptDated = metadata.resolveCatalogModelAcrossProvidersWithProvider('gpt-5.4-2026-04-01');
    expect(gptDated?.canonicalModelId).toBe('gpt-5.4');

    // Standalone image generation model.
    const gptImage = metadata.resolveCatalogModelAcrossProvidersWithProvider('gpt-image-2');
    expect(gptImage?.matchedProviderKind).toBe('openAI');
    expect(gptImage?.canonicalModelId).toBe('gpt-image-2');
    expect(gptImage?.metadata.profiles.imageGen).toBe('openaiImageGen');

    const claude = metadata.resolveCatalogModelAcrossProvidersWithProvider('claude-sonnet-4.5');
    expect(claude?.matchedProviderKind).toBe('anthropic');
    expect(claude?.metadata.displayName).toBe('Claude Sonnet 4.5');

    // Claude dated snapshot.
    const claudeDated = metadata.resolveCatalogModelAcrossProvidersWithProvider(
      'claude-sonnet-4-5-2026-04-01',
    );
    expect(claudeDated?.canonicalModelId).toBe('claude-sonnet-4.5');

    const gemini = metadata.resolveCatalogModelAcrossProvidersWithProvider('gemini-2.5-pro');
    expect(gemini?.matchedProviderKind).toBe('gemini');
    expect(gemini?.metadata.profiles.webSearch).toBe('geminiGoogleSearch');

    // Unknown model.
    expect(metadata.resolveCatalogModelAcrossProvidersWithProvider('my-custom-model')).toBeNull();
  });

  it('expands fixture generation parameters consistently: model-level enumValues override the platform shared schema, and absent ones fall back to the global values', async () => {
    const fixture = loadFixture();
    await seedMetadataCache({
      timestamp: Date.now(),
      data: fixture,
    });

    const metadata = await import('../metadata-client');
    await metadata.initMetadata();

    // gpt-5.4 declares model-level enumValues (["low", "high"]) that must override the four
    // global values from the platform shared schema, or the expert parameter panel would offer
    // levels the level picker deliberately narrowed away.
    const gpt = metadata.resolveCatalogModelAcrossProvidersWithProvider('gpt-5.4');
    const gptProfile = metadata.resolveGenerationProfileRef(gpt?.metadata.profiles.generation);
    expect(gptProfile?.parameters?.find((p) => p.id === 'reasoning_effort')?.enumValues).toEqual(['low', 'high']);

    // The reasoning_effort reference on claude-sonnet-4.5 declares no model-level enumValues, so
    // it falls back to the four global values of the platform shared schema (legacy wire path).
    const claude = metadata.resolveCatalogModelAcrossProvidersWithProvider('claude-sonnet-4.5');
    const claudeProfile = metadata.resolveGenerationProfileRef(claude?.metadata.profiles.generation);
    expect(claudeProfile?.parameters?.find((p) => p.id === 'reasoning_effort')?.enumValues).toEqual(['low', 'medium', 'high', 'xhigh']);
  });

  it('matches the relayRuntimeConfig snapshot in the fixture with the fallback defaults', async () => {
    const fixture = loadFixture();
    await seedMetadataCache({
      timestamp: Date.now(),
      data: fixture,
    });

    const metadata = await import('../metadata-client');
    await metadata.initMetadata();

    const runtime = metadata.getRelayRuntimeConfig();
    // The fixture version is a value frozen by the backend and should be passed through unchanged.
    expect(runtime.version).toBe('2026-04-23-1');
    expect(runtime.officialProviderWhitelist).toEqual([
      'openAI', 'anthropic', 'gemini', 'deepseek', 'miniMax', 'zhipu', 'qwen',
    ]);
    // The four transport defaults.
    expect(runtime.transportEnvelopes.openai_responses).toEqual({
      image: true, nativeFile: true, textFileInline: true,
      webSearch: true, imageGeneration: true, reasoning: true,
    });
    expect(runtime.transportEnvelopes.openai_chat_completions).toEqual({
      image: true, nativeFile: false, textFileInline: true,
      webSearch: false, imageGeneration: false, reasoning: true,
    });
    expect(runtime.transportEnvelopes.anthropic_messages).toEqual({
      image: true, nativeFile: true, textFileInline: true,
      webSearch: false, imageGeneration: false, reasoning: true,
    });
    expect(runtime.transportEnvelopes.gemini_generate_content).toEqual({
      image: true, nativeFile: true, textFileInline: true,
      webSearch: true, imageGeneration: true, reasoning: true,
    });
    expect(runtime.verificationPolicy).toEqual({
      hardFailedExpiryDays: 7,
      softFailedRetryAfterSeconds: 60,
      verifiedCacheDays: 30,
    });
    expect(runtime.featureGatingPolicy).toEqual({
      showActualModelIdHint: true,
      showSoftFailHint: true,
    });
  });

  it('falls back to the built-in default when relayRuntimeConfig.officialProviderWhitelist is an empty array, guarding against a malformed backend payload', async () => {
    const fixture = loadFixture() as Record<string, unknown> & {
      relayRuntimeConfig: Record<string, unknown>;
    };
    fixture.relayRuntimeConfig = {
      ...fixture.relayRuntimeConfig,
      officialProviderWhitelist: [],
    };
    await seedMetadataCache({
      timestamp: Date.now(),
      data: fixture,
    });

    const metadata = await import('../metadata-client');
    await metadata.initMetadata();

    const runtime = metadata.getRelayRuntimeConfig();
    // An empty array must be read as 'not sent' and fall back to the seven official providers.
    expect(runtime.officialProviderWhitelist).toHaveLength(7);
    expect(runtime.officialProviderWhitelist).toContain('openAI');
    expect(runtime.officialProviderWhitelist).toContain('anthropic');
  });

  it('uses the fallback for all four transports when relayRuntimeConfig.transportEnvelopes is an empty object', async () => {
    const fixture = loadFixture() as Record<string, unknown> & {
      relayRuntimeConfig: Record<string, unknown>;
    };
    fixture.relayRuntimeConfig = {
      ...fixture.relayRuntimeConfig,
      transportEnvelopes: {},
    };
    await seedMetadataCache({
      timestamp: Date.now(),
      data: fixture,
    });

    const metadata = await import('../metadata-client');
    await metadata.initMetadata();

    const runtime = metadata.getRelayRuntimeConfig();
    expect(runtime.transportEnvelopes.openai_responses.webSearch).toBe(true);
    expect(runtime.transportEnvelopes.openai_chat_completions.webSearch).toBe(false);
    expect(runtime.transportEnvelopes.anthropic_messages.imageGeneration).toBe(false);
    expect(runtime.transportEnvelopes.gemini_generate_content.webSearch).toBe(true);
  });
});

File: android/app/src/test/java/ai/oriveo/community/core/data/remote/MetadataFixtureTest.kt
package ai.oriveo.community.core.data.remote

import ai.oriveo.community.core.model.ProviderKind
import kotlinx.serialization.json.Json
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import kotlinx.serialization.KSerializer

/**
 * Android-side consumer test for the fixture shared across all three platforms.
 *
 * Reads `shared/test-fixtures/relay/metadata-fixture.json` and injects it into MetadataClient
 * via reflection, verifying that the fixture's hit samples, dated-snapshot aliases, and
 * relayRuntimeConfig fields line up with the Android fallback. Web and iOS have their own
 * equivalent tests.
 */
class MetadataFixtureTest {

    @After
    fun tearDown() {
        clearTable()
    }

    @Test
    fun `fixture covers gpt-5_4 snapshot alias and known official models`() {
        injectFixture()

        val gpt = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("gpt-5.4")
        assertNotNull(gpt)
        assertEquals(ProviderKind.OpenAI, gpt!!.matchedProviderKind)
        assertEquals("gpt-5.4", gpt.canonicalModelId)

        // dated-snapshot alias (a representative example)
        val gptDated = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("gpt-5.4-2026-04-01")
        assertNotNull(gptDated)
        assertEquals("gpt-5.4", gptDated!!.canonicalModelId)

        // a standalone image-generation model
        val gptImage = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("gpt-image-2")
        assertNotNull(gptImage)
        assertEquals(ProviderKind.OpenAI, gptImage!!.matchedProviderKind)
        assertEquals("gpt-image-2", gptImage.canonicalModelId)

        val claude = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("claude-sonnet-4.5")
        assertNotNull(claude)
        assertEquals(ProviderKind.Anthropic, claude!!.matchedProviderKind)

        val claudeDated = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("claude-sonnet-4-5-2026-04-01")
        assertNotNull(claudeDated)
        assertEquals("claude-sonnet-4.5", claudeDated!!.canonicalModelId)

        val gemini = MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("gemini-2.5-pro")
        assertNotNull(gemini)
        assertEquals(ProviderKind.Gemini, gemini!!.matchedProviderKind)

        assertNull(MetadataClient.resolveCatalogModelAcrossProvidersWithProvider("my-custom-model"))
    }

    @Test
    fun `fixture generation parameter expansion is consistent - model-level enumValues overrides shared schema, absence falls back to global`() {
        injectFixture()

        // gpt-5.4 declares model-level enumValues (["low", "high"]), which must override the four
        // global values from the shared platform schema -- otherwise the advanced parameter panel
        // would show a level picker with fake levels that were deliberately narrowed away.
        val gpt = MetadataClient.resolveCatalogModel("gpt-5.4", ProviderKind.OpenAI)
        val gptEnum = gpt?.profiles?.generation?.parameters
            ?.firstOrNull { it.id == "reasoning_effort" }
            ?.enumValues?.map { it.toString().trim('"') }
        assertEquals(listOf("low", "high"), gptEnum)

        // claude-sonnet-4.5's reasoning_effort reference declares no model-level enumValues, so it
        // must fall back to the four global values from the shared platform schema (the legacy
        // wire compatibility path).
        val claude = MetadataClient.resolveCatalogModel("claude-sonnet-4.5", ProviderKind.Anthropic)
        val claudeEnum = claude?.profiles?.generation?.parameters
            ?.firstOrNull { it.id == "reasoning_effort" }
            ?.enumValues?.map { it.toString().trim('"') }
        assertEquals(listOf("low", "medium", "high", "xhigh"), claudeEnum)
    }

    @Test
    fun `fixture relayRuntimeConfig aligns with android fallback defaults`() {
        injectFixture()

        val runtime = MetadataClient.relayRuntimeConfig()
        assertEquals("2026-04-23-1", runtime.version)
        assertEquals(
            listOf("openAI", "anthropic", "gemini", "deepseek", "miniMax", "zhipu", "qwen"),
            runtime.officialProviderWhitelist,
        )

        val openaiResponses = runtime.transportEnvelopes["openai_responses"]!!
        assertTrue(openaiResponses.image)
        assertTrue(openaiResponses.nativeFile)
        assertTrue(openaiResponses.webSearch)
        assertTrue(openaiResponses.imageGeneration)
        assertEquals("openAI", runtime.transportRules["openai_responses"]?.providerPriority)
        assertEquals("bearer", runtime.transportRules["openai_responses"]?.defaultAuthMode)
        assertEquals("v1", runtime.transportRules["openai_responses"]?.defaultVersion)
        assertEquals(listOf("v1"), runtime.transportRules["openai_responses"]?.acceptedVersions)
        assertEquals(true, runtime.transportRules["openai_responses"]?.codexIdentityDefault)
        assertEquals("web_search", runtime.transportRules["openai_responses"]?.webSearchToolName)
        assertEquals("inline_responses_tool", runtime.transportRules["openai_responses"]?.imageRoute)
        assertEquals(true, runtime.transportRules["openai_responses"]?.forceStreamForImageGeneration)

        val chat = runtime.transportEnvelopes["openai_chat_completions"]!!
        assertEquals(false, chat.nativeFile)
        assertEquals(false, chat.webSearch)
        assertEquals(false, chat.imageGeneration)
        assertEquals("bearer", runtime.transportRules["openai_chat_completions"]?.defaultAuthMode)
        assertEquals("images_endpoint", runtime.transportRules["openai_chat_completions"]?.imageRoute)

        val anthropic = runtime.transportEnvelopes["anthropic_messages"]!!
        assertEquals(false, anthropic.webSearch)
        assertEquals(false, anthropic.imageGeneration)
        assertEquals("anthropic", runtime.transportRules["anthropic_messages"]?.providerPriority)
        assertEquals("x_api_key", runtime.transportRules["anthropic_messages"]?.defaultAuthMode)

        val gemini = runtime.transportEnvelopes["gemini_generate_content"]!!
        assertTrue(gemini.webSearch)
        assertTrue(gemini.imageGeneration)
        assertEquals("x_goog_api_key", runtime.transportRules["gemini_generate_content"]?.defaultAuthMode)
        assertEquals("v1beta", runtime.transportRules["gemini_generate_content"]?.defaultVersion)

        assertEquals(7, runtime.verificationPolicy.hardFailedExpiryDays)
        assertEquals(60, runtime.verificationPolicy.softFailedRetryAfterSeconds)
        assertEquals(30, runtime.verificationPolicy.verifiedCacheDays)
        assertTrue(runtime.featureGatingPolicy.showActualModelIdHint)
        assertTrue(runtime.featureGatingPolicy.showSoftFailHint)
    }

    @Test
    fun `empty officialProviderWhitelist falls back to defaults`() {
        injectJson(
            """
            {
                "version": 1,
                "contractVersion": 1,
                "providers": {},
                "relayRuntimeConfig": {
                    "version": "2026-04-23-x",
                    "officialProviderWhitelist": [],
                    "transportEnvelopes": {},
                    "verificationPolicy": {
                        "hardFailedExpiryDays": 7,
                        "softFailedRetryAfterSeconds": 60,
                        "verifiedCacheDays": 30
                    },
                    "featureGatingPolicy": {
                        "showActualModelIdHint": true,
                        "showSoftFailHint": true
                    }
                }
            }
            """.trimIndent()
        )

        val runtime = MetadataClient.relayRuntimeConfig()
        // FALLBACK_RELAY_RUNTIME_CONFIG includes openAI / anthropic / gemini / deepseek / miniMax / zhipu / qwen / moonshot
        assertEquals(8, runtime.officialProviderWhitelist.size)
        assertTrue(runtime.officialProviderWhitelist.contains("openAI"))
        assertTrue(runtime.officialProviderWhitelist.contains("anthropic"))
        assertTrue(runtime.officialProviderWhitelist.contains("moonshot"))
    }

    @Test
    fun `empty transportEnvelopes falls back to defaults for all four transports`() {
        injectJson(
            """
            {
                "version": 1,
                "contractVersion": 1,
                "providers": {},
                "relayRuntimeConfig": {
                    "version": "2026-04-23-x",
                    "officialProviderWhitelist": ["openAI"],
                    "transportEnvelopes": {},
                    "verificationPolicy": {
                        "hardFailedExpiryDays": 7,
                        "softFailedRetryAfterSeconds": 60,
                        "verifiedCacheDays": 30
                    },
                    "featureGatingPolicy": {
                        "showActualModelIdHint": true,
                        "showSoftFailHint": true
                    }
                }
            }
            """.trimIndent()
        )

        val runtime = MetadataClient.relayRuntimeConfig()
        assertEquals(true, runtime.transportEnvelopes["openai_responses"]?.webSearch)
        assertEquals(false, runtime.transportEnvelopes["openai_chat_completions"]?.webSearch)
        assertEquals(false, runtime.transportEnvelopes["anthropic_messages"]?.imageGeneration)
        assertEquals(true, runtime.transportEnvelopes["gemini_generate_content"]?.webSearch)
    }

    private fun injectFixture() {
        // The gradle test task's cwd is the app module directory.
        //   .../android/app/
        //   -> parentFile walks back up to the repo root
        val moduleDir = File(System.getProperty("user.dir") ?: ".").absoluteFile
        val repoRoot = moduleDir.parentFile!!.parentFile!!
        val fixturePath = File(repoRoot, "shared/test-fixtures/relay/metadata-fixture.json")
        require(fixturePath.exists()) { "fixture not found at: ${fixturePath.absolutePath}" }
        injectJson(fixturePath.readText(Charsets.UTF_8))
    }

    private fun injectJson(rawJson: String) {
        val clientClass = MetadataClient::class.java

        val jsonField = clientClass.getDeclaredField("json")
        jsonField.isAccessible = true
        val internalJson = jsonField.get(MetadataClient.instance) as Json

        val responseClass = clientClass.declaredClasses.first { it.simpleName == "MetadataResponse" }
        val companionField = responseClass.getDeclaredField("Companion")
        companionField.isAccessible = true
        val companion = companionField.get(null)
        val serializerMethod = companion.javaClass.getDeclaredMethod("serializer")
        serializerMethod.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val serializer = serializerMethod.invoke(companion) as KSerializer<Any>

        val decoded = internalJson.decodeFromString(serializer, rawJson)

        val tableField = clientClass.getDeclaredField("table")
        tableField.isAccessible = true
        tableField.set(MetadataClient.instance, decoded)
    }

    private fun clearTable() {
        val tableField = MetadataClient::class.java.getDeclaredField("table")
        tableField.isAccessible = true
        tableField.set(MetadataClient.instance, null)
    }
}
